File: bot/models/product.py
"""
Product model
"""

from datetime import datetime, UTC
from .database import get_worksheet, add_record, update_record, find_record, get_all_records


class Product:
    def __init__(self, id=None, name=None, description=None, price=0.0, stock=0, 
                 is_available=True, created_at=None, updated_at=None):
        self.id = int(id) if id else None
        self.name = str(name) if name else ''
        self.description = str(description) if description else ''
        self.price = float(price) if price else 0.0
        self.stock = int(float(stock)) if stock else 0
        self.is_available = str(is_available).lower() == 'true' if isinstance(is_available, str) else bool(is_available)
        self.created_at = str(created_at) if created_at else datetime.now(UTC).isoformat()
        self.updated_at = str(updated_at) if updated_at else datetime.now(UTC).isoformat()
    
    def save(self):
        """Save product to Google Sheets"""
        data = {
            'name': self.name,
            'description': self.description,
            'price': self.price,
            'stock': self.stock,
            'is_available': str(self.is_available),
            'created_at': self.created_at,
            'updated_at': datetime.now(UTC).isoformat()
        }
        
        if self.id:
            data['id'] = self.id
            update_record('Products', self.id, data)
        else:
            result = add_record('Products', data)
            self.id = result['id']
        return self
    
    @staticmethod
    def get_by_id(product_id):
        """Get product by ID"""
        _, record = find_record('Products', 'id', product_id)
        if record:
            return Product(**record)
        return None
    
    @staticmethod
    def get_all(available_only=False):
        """Get all products"""
        records = get_all_records('Products')
        products = [Product(**record) for record in records]
        if available_only:
            products = [p for p in products if p.is_available]
        return products
    
    def __repr__(self):
        return f"<Product(id={self.id}, name='{self.name}', price={self.price})>"
    
    def to_dict(self):
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'price': self.price,
            'stock': self.stock,
            'is_available': self.is_available
        }

File: src/models/product.ts
/**
 * Product model
 */

import { addRecord, updateRecord, findRecord, getAllRecords } from './database';

export interface IProduct {
  id?: number;
  name: string;
  description: string;
  price: number;
  stock: number;
  is_available: boolean | string;
  created_at?: string;
  updated_at?: string;
}

export class Product {
  id?: number;
  name: string;
  description: string;
  price: number;
  stock: number;
  is_available: boolean;
  created_at: string;
  updated_at: string;

  constructor(data: IProduct) {
    this.id = data.id ? parseInt(String(data.id)) : undefined;
    this.name = String(data.name || '');
    this.description = String(data.description || '');
    this.price = parseFloat(String(data.price || 0));
    this.stock = parseInt(String(data.stock || 0));
    
    // Handle boolean conversion for is_available
    if (typeof data.is_available === 'string') {
      this.is_available = data.is_available.toLowerCase() === 'true';
    } else {
      this.is_available = Boolean(data.is_available !== false);
    }
    
    this.created_at = data.created_at || new Date().toISOString();
    this.updated_at = data.updated_at || new Date().toISOString();
  }

  /**
   * Save product to Google Sheets
   */
  async save(): Promise<Product> {
    const data = {
      name: this.name,
      description: this.description,
      price: this.price,
      stock: this.stock,
      is_available: String(this.is_available),
      created_at: this.created_at,
      updated_at: new Date().toISOString()
    };

    if (this.id) {
      await updateRecord('Products', this.id, { id: this.id, ...data });
    } else {
      const result = await addRecord('Products', data);
      this.id = result.id;
    }

    return this;
  }

  /**
   * Get product by ID
   */
  static async getById(productId: number): Promise<Product | null> {
    const [, record] = await findRecord('Products', 'id', productId);
    return record ? new Product(record as IProduct) : null;
  }

  /**
   * Get all products
   */
  static async getAll(availableOnly: boolean = false): Promise<Product[]> {
    const records = await getAllRecords('Products');
    let products = records.map(record => new Product(record as IProduct));

    if (availableOnly) {
      products = products.filter(p => p.is_available);
    }

    return products;
  }

  /**
   * Convert to plain object
   */
  toDict(): IProduct {
    return {
      id: this.id,
      name: this.name,
      description: this.description,
      price: this.price,
      stock: this.stock,
      is_available: this.is_available
    };
  }
}
